### DISCUSSION TITLE
How to terminate @work worker on app exit?

### DISCUSSION THREAD

**Discussion by jehon** [Textualize/textual#2839] [Q&A] [answered]
I have a widget with a @work worker:

```
    @work(exclusive=True)
    def start(self) -> None:
          blablabla
```

When exiting the application (app.exit()), python wait for my worker to finish...

1/ How could I kill that thread/worker? (stronger than cancel?)
2/ Furher: could that be done from inside a widget? I mean, is there a function called when the app is exiting?

**Comment by TomJGooding** [comment, marked as answer]
I assumed that any workers would be cancelled on exiting the app? According to the docs for  [Worker lifetimes](https://textual.textualize.io/guide/workers/#worker-lifetime):

> Workers are tied to the DOM node (widget, screen, or app) where they are created. This means that if you remove the widget or pop the screen where they are created, then the tasks will be cleaned up automatically. Similarly **if you exit the app, any running tasks will be cancelled.** [emphasis mine]

[EDIT: Sorry I've just spotted your example doesn't have the `async` keyword, so would actually create a [thread worker](https://textual.textualize.io/guide/workers/#thread-workers). Just to double-check, is it specifically threaded code you are asking about?]

**Reply by jehon** [reply]
Using worker and async code was a better idea. Now, with @work async blablabla, the code work as expected.

Thanks for pointing me in the correct direction
